Feature/child rows

interim17 opened the pull request:
Time Estimate or Size
=======
_Medium_

Problem
=======
We need to conditionally render rows of checkboxes/color swatch/name for an agent's display states.

Hopefully this is the final bit of agent selection functional work, prior to a styling PR.
Open to feedback/refactoring to move this along., there is definitely more than one way to do this feature.

Solution
========
In `AgentRow` define a function to map over displayStates called `getDisplayStateRows`

In `types` you'll see a union type for the different props a `Checkbox` can take, the purpose of this union type is to reduce the number of checks we need on possibly undefined props, and make it clear whether we are dealing with a top level agent, or a display state, without a lot of fuss. In turn we can use the same Hide and Highlight checkbox components for both levels of organizaton.

I still have only those two basic checkboxes, and two functions in Context for handling when they are clicked:
`handleVisibilityCheckboxChange` and `handleHightlightCheckboxChange`.

I find this collects the logic and makes it readable, and I don't think it's overly generalized. Functions to get checkbox state still live in the components.

I also updated tests.

## Type of change
* New feature (non-breaking change which adds functionality)

Steps to Verify:
----------------
1. Run the trajectory and click the button to show the display state rows.
2. Try the various checkbox combinations.

Screenshots (optional):
-----------------------
![Screenshot 2024-04-11 at 4 19 28 PM](https://github.com/simularium/nbsv/assets/24981838/4e393e1d-36d1-45c8-9e54-624f9257511c)

interim17 committed:
add checkbox types

interim17 committed:
render rows for display states and their checkboxes

interim17 committed:
update tests

interim17 commented on src/AgentVisibilityContext.tsx:
```diff
@@ -37,16 +47,54 @@ export const VisibilityProvider = ({ children }: { children: ReactNode }) => {
-    setHiddenAgents((prevHiddenAgents) =>
-      getNewSelectionMap(agentName, prevHiddenAgents)
-    );
+  const getDisplayStates = (agentName: string): string[] => {
+    const agentEntry = uiDisplayData.find((entry) => entry.name === agentName);
+    if (agentEntry === undefined) {
+      return [];
+    }
+    return agentEntry.displayStates.map((state) => state.name);
```
When clicking on a checkbox we may need to put its siblings names into selection, hence this function.

interim17 commented on src/types.ts:
```diff
@@ -25,3 +25,25 @@ export enum SelectionType {
+  agentType: AgentType;
+}
+
+export interface AgentProps extends BaseCustomCheckboxProps {
+  agentType: AgentType.Agent;
+  displayStateName?: never;
+}
+
+export interface DisplayStateProps extends BaseCustomCheckboxProps {
+  agentType: AgentType.DisplayState;
+  displayStateName: string;
+}
```
These two types enforce that an Agent Checkbox may never have display state name, and a display state Checkbox must always have one.

interim17 commented on src/AgentVisibilityContext.tsx:
```diff
@@ -37,16 +47,54 @@ export const VisibilityProvider = ({ children }: { children: ReactNode }) => {
+    }
+  };
+
+  const handleHightlightCheckboxChange = (
+    agentName: string,
+    displayStateName?: string
+  ) => {
+    if (displayStateName !== undefined) {
+      const displayStates = getDisplayStates(agentName);
+      setHighlightedAgents((prevHiddenAgents) =>
+        getNewMapAfterDisplayStateClick(
+          agentName,
+          displayStateName,
+          displayStates,
+          prevHiddenAgents
+        )
+      );
+    } else {
+      setHighlightedAgents((prevHiddenAgents) =>
+        getNewMapAfterTopLevelCheckboxClick(agentName, prevHiddenAgents)
+      );
+    }
```
Maintained the same two handlers from the last change set here and generalized them to work for both levels of Checkbox. Splitting them up means moving the conditional check into the component, or making new types of checkboxes. Both are doable, this was my preference. Thoughts?

ShrimpCryptid commented on src/utils.ts:
```diff
@@ -1,15 +1,54 @@
+  return newMap;
+};
+
+export const getNewMapAfterDisplayStateClick = (
```
The name DisplayState doesn't really suggest the parent/child relationship between agents and display states. Is there a better name that would make this more clear?

ShrimpCryptid replied:
Could we just call them `agentChildDisplayState`?

interim17 replied:
I agree that `displayState` name is clunky, but that typing is defined in the viewer so I think it's out of scope for this PR, but it's definitely one of the things about the `UiDisplayData` data structure that could be improved.

My reasoning behind sticking with that naming convention in the front end, where we could just substitute 'child' here in the function name, (`getNewMapAfterChildCheckboxClick` for example) was to keep explicit what data structures we are using to arrive at these components, because there is a lot of converting back and forth and use of different terminology that makes harder to follow what is derived from what (hide, highlight, visibility, selection or: agent, displayState, topLevel, shared, child, etc.)

ShrimpCryptid commented on src/utils.ts:
```diff
@@ -1,15 +1,54 @@
+  const newMap: VisibilitySelectionMap = { ...currentVisibilityMap };
+  const currentStates = currentVisibilityMap[agentName] || [];
+  const index = currentStates.indexOf(displayStateName);
+  if (currentStates.length === 0) {
```
```suggestion
  if (currentStates.length === 0) {
    // Agent + all of its child display states are currently selected, so update the selected states
    // to include all of the children except for `displayStateName`.
```

Suggested comment here to explain what this is doing.

GENERAL THOUGHT: I don't love that we're trying to pack extra information into what's inside of the `currentStates` list based on whether it's empty or the agent name... It makes it much harder to read this code.

I kind of want this behavior to be controlled by an enum. Something like, `Selected.ALL`, `Selected.PARTIAL`, `Selected.NONE`, and then some payload that's included with `Selected.PARTIAL`...?

```
const enum Selected {
  PARTIAL,
  NONE,
  ALL
}

type SelectionStatus = { selected: Selected.NONE | Selected.ALL } & { selected: Selected.PARTIAL, selectedAgentChildren: string[] }
```

ShrimpCryptid replied:
Alternatively, making a human-readable method name to handle these checks would make this code easier to parse.

For example:
```
const isNoAgentChildrenSelected= (agentName: string, selectedChildren: string[]): boolean => { return selectedChildren.length === 1 && selectedChildren[0] === agentName; }

// Potentially unsafe because you could have every child in the `selectedChildren` array and it would still think that not all children are selected. 
const isAllAgentChildrenSelected = (_agentName: string, selectedChildren: string[]): boolean => { return selectedChildren.length === 0; }

const isAgentPartiallySelected = (agentName: string, selectedChildren: string[]): boolean => { return !isNoAgentChildrenSelected(agentName, selectedChildren) && !isAllAgentChildrenSelected (agentName, selectedChildren); }

if (isAllAgentChildrenSelected (agentName, currentStates)) {
 ...
} else {
  if (isNoAgentChildrenSelected(agentName, currentStates)) {
    ...
  } else {
    ...
  }
}
```

interim17 replied:
To your general thought: I completely agree. When there is less time pressure on review I think we should have an intermediate data structure that sort of enhances/replaces `UiDisplayData` and does the work of `VisibilitySelectionMap` without the arbitrariness of the empty arrays and arrays with agent names in them.

Either way I agree that some of these changes to function names and use of enums can make this more readable and I'll start on that.

ShrimpCryptid commented on src/utils.ts:
```diff
@@ -1,15 +1,54 @@
+  agentName: string,
+  displayStateName: string,
+  allDisplayStates: string[],
+  currentVisibilityMap: VisibilitySelectionMap
```
`VisibilitySelectionMap` should really reference what it's mapping to and from as a type.

`AgentsToSelectedChildren` or `AgentsToVisibilitySelection`?

interim17 replied:
Agreed, I think `AgentsToVisibilitySelection` is better.... Is this another scope question? I'm using the same `VisibilitySelectionMap` as the website. Would it be confusing to have the same data structure named two different ways in NBSV and website? Might want some buy in on a name change from @meganrm and update it in both places...

ShrimpCryptid commented on src/utils.ts:
```diff
@@ -1,15 +1,54 @@
+): VisibilitySelectionMap => {
+  const newMap: VisibilitySelectionMap = { ...currentVisibilityMap };
+  const currentStates = currentVisibilityMap[agentName] || [];
+  const index = currentStates.indexOf(displayStateName);
```
Index is only used once; would be more readable to move this check directly into the `if` branch that evaluates its value

ShrimpCryptid commented on src/utils.ts:
```diff
@@ -1,15 +1,54 @@
+  agentName: string,
+  displayStateName: string,
+  allDisplayStates: string[],
+  currentVisibilityMap: VisibilitySelectionMap
```
If we're not able to refactor or rename it, would really appreciate a comment on `VisibilitySelectionMap` that describes how to parse it at the very least!

I didn't really understand it until you talked me through it, and there isn't really any documentation for it in this codebase.
> if visMap[agentName] = [agentName] ... the agent is not selected (Default)
> if visMap[agentName] = [ ] ... the agent is selected and by extension any display states are also selected
> if visMap[agentName] = [ displayStateName1, displayStateName2... ] .. some but not all display states are selected, the agent checkbox is by definition indeterminate

interim17 replied:
Where do you think would be the most useful place for this comment, next to the type definition?

ShrimpCryptid replied:
Next to the type definition makes sense to me, since that was the first place I checked.

ShrimpCryptid reviewed (commented):
Going to leave this initial round of comments all focused on `utils`. I find `VisibilitySelectionMap` really hard to read and understand...

interim17 committed:
dont declare const for index value that is only used once

interim17 committed:
add comment to explain use of visibility seleciton map

interim17 committed:
use semantic constants and functions to check selection status

interim17 commented:
@ShrimpCryptid I put a comment by the type, and added semantic constants and functions for the conditional checks in `getNewMapAfterDisplayStateClick`

I think there is more to do to improve this part of the code base here and in the website but this is my best attempt to clean it up without scope creep, lmk if this is more readable?

meganrm commented on src/components/AgentRow.tsx:
```diff
@@ -13,19 +15,67 @@ const AgentRow: React.FC<AgentRowProps> = (
 ): JSX.Element => {
   const { agent } = props;
 
+  const [showDisplayStates, setShowDisplayStates] = React.useState(false);
```
in the website all mention of "displayState" is confined to the 'state'. But in the UI components, the language is `parent/child`. The idea of display states comes from our data structure and the semantics of how the trajectories are organized. But I think at this level in the code, `showChildren` and `hasChildren` is a lot easier to understand

interim17 replied:
I agree it’s easier to understand in this context and I’m willing to change this! I also found it confusing that this same data has a few different names throughout different levels of the code so I thought we could reduce effort in understanding how parts of the code base connect by using the same nomenclature in more places.

meganrm commented on src/types.ts:
```diff
@@ -25,3 +25,25 @@ export enum SelectionType {
+
+export enum AgentType {
+  Agent,
+  DisplayState,
```
this is a little confusing because the viewer also has `agentType` which means something really different. I think something like a util that tells `hasChildren` or `isChild` would be more clear

interim17 replied:
Does it mean something really different? I thought agent type and tags were basically the way that agent/display state were described when decoded from sim file format.

I’m definitely open to using the hasChildren isChild language

meganrm commented on src/types.ts:
```diff
@@ -25,3 +25,25 @@ export enum SelectionType {
+
+export interface AgentProps extends BaseCustomCheckboxProps {
+  agentType: AgentType.Agent;
+  displayStateName?: never;
```
why can't this be a string?

interim17 replied:
The two enum members? It could be. I’ve been given the feedback when using strings in this way to formalize them and use an enum..

interim17 committed:
rename display state language to use parent/child and rename agent type to agent level

interim17 commented:
I've changed all language of display states to children and renamed `AgentType` to `AgentLevel` and it's members to `Agent` and `ChildAgent`. 

Not wedded to that language, the purpose of that type is to keep the code DRY and let us distinguish easily between the two levels of organization when using checkboxes, I don't feel a strong opinion about what it's called.

meganrm commented on src/AgentVisibilityContext.tsx:
```diff
@@ -37,16 +47,54 @@ export const VisibilityProvider = ({ children }: { children: ReactNode }) => {
-    );
+  const handleVisibilityCheckboxChange = (
+    agentName: string,
+    childAgentName?: string
```
sorry to be so picky about this, but this also doesn't really make sense. the children are different states the agent can be in. It's not another agent name. So this second thing only is defined if you click on a child checkbox right? I might reverse this, have the first argument just be the name of the checkbox, and have an optional `parentName` if it's a child checkbox.

meganrm commented on src/utils.ts:
```diff
@@ -1,15 +1,61 @@
+export const getNewMapAfterChildAgentClick = (
+  agentName: string,
+  childAgentName: string,
+  allChildren: string[],
```
why does this function need both the agentName and allChildren? Isn't it able to get the children by the agent name?

interim17 replied:
I pulled these functions out of the context file in response to feedback, so they lost access to the context as they are not in the component tree and not wrapped by the provider. That has driven some of the way this all gets written because the utils/selectors only know what is passed in to them.

meganrm replied:
the one above does seem like a util function because it's just mapping over data and then returning the same thing every time. I see why you wanted these functions next to each other, but this one is not really like that one. There are a bunch of functions within this function that could be considered util functions

meganrm commented on src/utils.ts:
```diff
@@ -1,15 +1,61 @@
+    }
+    if (allChildrenNowSelected()) {
+      newMap[agentName] = [];
+    }
```
at first glance, this function is really hard to read. and it also no longer feels like a util function because it's making a lot of decisions based on the data (not just transforming one data structure to another)

meganrm commented on src/components/HighlightCheckbox.tsx:
```diff
@@ -1,40 +1,45 @@
-  const getHighlightCheckboxStatus = (): CheckboxState => {
-    if (highlightedAgents[agent.name]?.length === 0) {
+  const selections = highlightedAgents[agentName];
+  const isTopLevel = childName === undefined;
```
the fact that you have to call different functions for a different type makes me feel like topLevel should just be a different component from a child/regular checkbox. this may just be a personal preference, curious what you think @ShrimpCryptid. Like a ` CheckBox.Level === Parent ? <TopLevelCheckbox /> : <Checkbox />`

ShrimpCryptid replied:
No strong opinions on splitting it up, though there may be some repeated code that would be bad for DRY. I do agree it's a little funny having two different functions. Can they be merged?

meganrm commented on src/components/HighlightCheckbox.tsx:
```diff
@@ -1,40 +1,45 @@
+  const isTopLevel = childName === undefined;
+
+  const determineTopLevelCheckboxStatus = () => {
+    if (selections?.length === 0) {
```
why is this possibly undefined?

meganrm commented on src/components/HideCheckbox.tsx:
```diff
@@ -1,47 +1,48 @@
-  const getHideCheckboxStatus = (): CheckboxState => {
-    if (hiddenAgents[agent.name]?.length === 0) {
+  const selections = hiddenAgents[agentName];
+  const isTopLevel = childName === undefined;
```
if you have the `AgentLevel` why are you doing this check here?

meganrm commented on src/components/AgentRow.tsx:
```diff
@@ -13,19 +15,67 @@ const AgentRow: React.FC<AgentRowProps> = (
+      const checkboxProps: ChildAgentProps = {
+        agentLevel: AgentLevel.ChildAgent,
+        agentName: agent.name,
+        childName: displayState.name,
```
this feels like the level where all the 'is selected' stuff should be happening. you could even define the handler here. That way the props sent into a checkbox could jsut be its value/name, whether it's checked or not, and an onClick handler

meganrm commented on src/types.ts:
```diff
@@ -25,3 +25,25 @@ export enum SelectionType {
+
+export interface ParentAgentProps extends BaseCustomCheckboxProps {
+  agentLevel: AgentLevel.Agent;
+  childName?: never;
```
With a previous comment what I meant was why isn't this `childName?: string`

meganrm replied:
I'm confused about this again: shouldn't a parent have the possibility of multiple children? so shouldn't this be `children?: string[]`?  or `childrenNames?: string[];

meganrm commented on src/types.ts:
```diff
@@ -25,3 +25,25 @@ export enum SelectionType {
   Highlight,
 }
+
+export enum AgentLevel {
```
this is more `checkBoxLevel` than agent level. Sorry if this was unclear, but agents aren't ever children. Checkboxes have children, agents have states

meganrm commented on src/types.ts:
```diff
@@ -25,3 +25,25 @@ export enum SelectionType {
+  ChildAgent,
+}
+
+interface BaseCustomCheckboxProps {
+  agentName: string;
+  agentLevel: AgentLevel;
+}
```
```suggestion
interface BaseCustomCheckboxProps {
  name: string;
  level: CheckboxLevel;
}
```

meganrm commented on src/types.ts:
```diff
@@ -25,3 +25,25 @@ export enum SelectionType {
+  agentLevel: AgentLevel;
+}
+
+export interface ParentAgentProps extends BaseCustomCheckboxProps {
+  agentLevel: AgentLevel.Agent;
+  childName?: never;
+}
```
```suggestion
export interface ParentAgentProps extends BaseCustomCheckboxProps {
  agentLevel: AgentLevel.Agent;
  parentName: string;
}
```

meganrm commented on src/types.ts:
```diff
@@ -25,3 +25,25 @@ export enum SelectionType {
   Highlight,
 }
+
+export enum AgentLevel {
+  Agent,
+  ChildAgent,
+}
```
```suggestion
export enum Level {
  Parent,
  Child,
}
```

meganrm commented on src/types.ts:
```diff
@@ -25,3 +25,25 @@ export enum SelectionType {
+  childName?: never;
+}
+
+export interface ChildAgentProps extends BaseCustomCheckboxProps {
+  agentLevel: AgentLevel.ChildAgent;
+  childName: string;
+}
```
```suggestion
export interface ChildCheckboxProps extends BaseCustomCheckboxProps {
  level: Level.Child;
  parentName: string;
}
```

then the main name would be the child name. From the point of view of this node, its name is just its name. It has a parent.

ShrimpCryptid commented on src/components/HighlightCheckbox.tsx:
```diff
@@ -1,40 +1,45 @@
+  const selections = highlightedAgents[agentName];
+  const isTopLevel = childName === undefined;
+
+  const determineTopLevelCheckboxStatus = () => {
```
Tiny nit: `get` instead of `determine`

interim17 committed:
use UserChangesMap to track user selections and split checkboxes by parent/child level

interim17 committed:
update tests

interim17 committed:
make default state of userChangesMap value an empty array and add agents to array when selected

interim17 committed:
rename initializer for uidata getInitialUserSelections

ShrimpCryptid commented on src/AgentVisibilityContext.tsx:
```diff
@@ -30,71 +26,48 @@ export const VisibilityContext = createContext<VisibilityContextType>({
-    const noAgentsSelectedMap = mapUIDisplayDataToSelectionMap(data);
-    setHighlightedAgents(noAgentsSelectedMap);
-    setHiddenAgents(noAgentsSelectedMap);
+    const initialSelections = getInitialUserSelections(data);
```
Nit: `makeEmptyUserSelections`? `getInitialUserSelections` suggests that you're retrieving some kind of default to me.

interim17 replied:
oh boy, went back and forth with @meganrm a lot on this one, kind of scared to change it again at this point haha, Megan, thoughts?

I like both options here, just need one everyone can agree on.

meganrm replied:
yeah I like makeEmptyUserSelections

interim17 replied:
@meganrm then the people have spoken

ShrimpCryptid commented on src/AgentVisibilityContext.tsx:
```diff
@@ -30,71 +26,48 @@ export const VisibilityContext = createContext<VisibilityContextType>({
-    setHighlightedAgents(noAgentsSelectedMap);
-    setHiddenAgents(noAgentsSelectedMap);
+    const initialSelections = getInitialUserSelections(data);
+    setHighlightedAgents(initialSelections);
+    setHiddenAgents(initialSelections);
```
```suggestion
    setHighlightedAgents(getInitialUserSelections(data));
    setHiddenAgents(getInitialUserSelections(data));
```

Maybe worth calling `getInitialUserSelections` twice, once for each of these setters, so they aren't sharing the same object reference? (I think it's fine based on how you're setting values, but it makes it clear that these aren't supposed to be sharing a map reference.)

interim17 replied:
Yeah all the updates are functionalized so it shouldn't be an issue, and my default is to store something if you're going to use the same value twice?

But maybe we can split the different on the two suggestions here and keep the function name and drop the shared variable. I'll push that change.

ShrimpCryptid commented on src/AgentVisibilityContext.tsx:
```diff
@@ -30,71 +26,48 @@ export const VisibilityContext = createContext<VisibilityContextType>({
+  const handleHideChildCheckboxChange = (name: string, parent: string) => {
+    setHiddenAgents((prevHiddenAgents) =>
+      getSelectionAfterChildCheckboxClick(name, parent, prevHiddenAgents)
+    );
```
Really good readability change.

ShrimpCryptid commented on src/constants.ts:
```diff
@@ -26,15 +26,14 @@ export const VIEWER_INITIAL_WIDTH: number = 500;
- * if visMap[agentName] = [agentName] // the agent is not selected (default), and by extension, no children are selected
- * if visMap[agentName] = [ ] // the agent is selected, and by extension all children are selected
- * if visMap[agentName] = [ childAgentName1, childAgentName2... ] // some but not all children are selected, the agent checkbox is by definition indeterminate
- * if a selection array would contain all the child names, it should be changed to an empty array
```
I'd like to bring back the section here that explains how to think about values of the `UserChangesMap`. For example, what the the default state of including the agent name mean?

ShrimpCryptid commented on src/components/ChildHideCheckbox.tsx:
```diff
@@ -0,0 +1,37 @@
+  const tooltipText = tooltipMap[checkboxStatus];
+
+  return (
+    <Tooltip placement="right" title={tooltipText}>
```
```suggestion
    <Tooltip placement="right" title={tooltipText} trigger=["focus", "hover"]>
```
For this and other tooltips, consider making them responsive to keyboard focusing. It might also be nice to add a visually hidden label?

interim17 replied:
Will add this change! Focus states inside jupyer notebooks is a topic I'm scared to open at the moment but this is a good change for now.

ShrimpCryptid commented on src/components/ChildHighlightCheckbox.tsx:
```diff
@@ -0,0 +1,77 @@
+    getHighlightDisplayOptions(checkboxStatus);
+
+  return (
+    <Tooltip placement="top" title={tooltipText}>
```
See previous comment about focus triggers

ShrimpCryptid commented on src/__tests__/utils.spec.ts:
```diff
@@ -1,270 +1,184 @@
-        'agent3',
-        mockVisibilitySelectionMap
+  describe('updateUserChangesAfterCheckboxClick', () => {
+    it('removes an agent with no children from selection if it was previoulsy included', () => {
```
```suggestion
    it('removes an agent with no children from selection if it was previously included', () => {
```

ShrimpCryptid commented on src/__tests__/utils.spec.ts:
```diff
@@ -1,270 +1,184 @@
-    it('returns an array of SelectionEntry objects', () => {
-      const payload = convertMapToSelectionStateInfo(
-        mockVisibilitySelectionMap
+    it('should return a selection entry with an empty tags array if agent name is in user changes map has agent name in array', () => {
```
```suggestion
    it('should return a selection entry with an empty tags array if agent name is in user changes map', () => {
```

interim17 replied:
will fix the typo (adding "and") but actually need that last phrase for accuracy

ShrimpCryptid commented on src/__tests__/utils.spec.ts:
```diff
@@ -1,270 +1,184 @@
-          name: 'agent2',
-          tags: [],
-        },
+    it('should return no selection entry for an agent withtout children if the user changes map array is empty', () => {
```
```suggestion
    it('should return no selection entry for an agent without children if the user changes map array is empty', () => {
```

ShrimpCryptid reviewed (commented):
Left many minor comments, this PR is getting really close! Once you address those I'm happy to leave a final approval.

interim17 committed:
Update src/__tests__/utils.spec.ts

Co-authored-by: Peyton Lee <peyton.lee@alleninstitute.org>

interim17 committed:
Update src/__tests__/utils.spec.ts

Co-authored-by: Peyton Lee <peyton.lee@alleninstitute.org>

interim17 committed:
Update src/components/ChildHideCheckbox.tsx

Co-authored-by: Peyton Lee <peyton.lee@alleninstitute.org>

interim17 committed:
fix typo in test

interim17 committed:
dont use shared variable for setting initial hidden and highlighted agetns

interim17 committed:
add focus trigger to tooltips

ShrimpCryptid reviewed (approved):
LGTM! Thank you for the hard work on this :')

interim17 committed:
rename util to makeEmptyUserSelections

interim17 committed:
flesh out comment on userChangesMap

meganrm commented on src/__tests__/utils.spec.ts:
```diff
@@ -1,107 +1,184 @@
+      const result = updateUserChangesAfterCheckboxClick(
+        'agent_with_no_children',
+        [],
+        {
```
nit: it would be nice to define this block in all these tests as `const prevState= {...}` and then send into updateUserChangesAfterCheckboxClick(name, prevState) just for readability

meganrm commented on src/__tests__/utils.spec.ts:
```diff
@@ -1,107 +1,184 @@
+        agent_with_children: [],
+      });
+    });
+    it('selects agents children if they were not previously selected', () => {
```
nit: the other test descriptions day 'remove/add' this says 'select'

meganrm commented on src/components/ChildHideCheckbox.tsx:
```diff
@@ -0,0 +1,37 @@
+  const selections = hiddenAgents[parentName];
+
+  const getCheckboxStatus = () => {
+    if (selections?.includes(name)) {
```
I asked this before, but why is selections possibly undefined? shouldn't it always be an array?

interim17 replied:
Selections is specific to the agent, so in the initial state before the UI display data arrives this would be undefined.

meganrm commented on src/components/ChildHideCheckbox.tsx:
```diff
@@ -0,0 +1,37 @@
+  const { handleHideChildCheckboxChange, hiddenAgents } =
+    useContext(VisibilityContext);
+
+  const selections = hiddenAgents[parentName];
```
```suggestion
// the selections are initialized with the parents names as keys, 
//so this will always exist, but typeScript doesn't know that. 
   const selections = hiddenAgents[parentName] || [];
```
(probably have to do some formatting on the comment to accept this change)

interim17 replied:
ok ill do it manually and add it in the highlighted one as well

meganrm commented on src/components/HideCheckbox.tsx:
```diff
@@ -1,47 +1,42 @@
 
-  const getHideCheckboxStatus = (): CheckboxState => {
-    if (hiddenAgents[agent.name]?.length === 0) {
+  const selections = hiddenAgents[agent.name];
```
it feels to me like selections should be a prop of all these components. You could also make the `handleHideCheckboxChange(agent.name, children` at the higher level and then this component wouldn't have to access the context at all. but it's just a preference, not a blocking issue. My thinking is that you access the context once in `AgentRow` and then send all the necessary information into each of the subcomponents since they're already receiving some props. but if you think it's clearer this way it's fine

meganrm commented on src/constants.ts:
```diff
@@ -24,7 +24,19 @@ export const SIDE_PANEL_WIDTH: number = 280;
+/**
+ * The UserChangesMap stores user changes to highlighting and hiding checkboxes.
+ * If a user highlights an agent or display state, that "star" checkbox will be checked.
+ * If a user hides an agent or display state, that checkbox will be checked.
```
you mean unchecked right? but included in the UserChangesMap

meganrm commented on src/constants.ts:
```diff
@@ -24,7 +24,19 @@ export const SIDE_PANEL_WIDTH: number = 280;
+ * If a user highlights an agent or display state, that "star" checkbox will be checked.
+ * If a user hides an agent or display state, that checkbox will be checked.
+ * The selection state info for the viewer is derived from these maps.
+ * The default state is to include the agent name, or if an agent has multiple display states,
```
this is backwards too, your default state is to include nothing

meganrm commented on src/constants.ts:
```diff
@@ -24,7 +24,19 @@ export const SIDE_PANEL_WIDTH: number = 280;
+ * The selection state info for the viewer is derived from these maps.
+ * The default state is to include the agent name, or if an agent has multiple display states,
+ * all of the display state names.
+ * userChangesMap[agentName] = [] means nothing is selected
```
means nothing is highlighted or nothing is hidden

meganrm commented on src/constants.ts:
```diff
@@ -24,7 +24,19 @@ export const SIDE_PANEL_WIDTH: number = 280;
+ * userChangesMap[agentName] = [] means nothing is selected
+ * userChangesMap[agentName] = [agentName] means the agent is selected in a case with no display states
+ * userChangesMap[agentName] = [displayState1, displayState2] means whatever display states are in the array are selected
+ */
```
I think this whole thing needs to be looked at again

interim17 committed:
update userchanges map comment

meganrm commented on src/constants.ts:
```diff
@@ -24,7 +24,19 @@ export const SIDE_PANEL_WIDTH: number = 280;
-export interface VisibilitySelectionMap {
+
+/**
+ * The UserChangesMap stores user changes to highlighting and hiding (made by clicking checkboxes).
```
nit: `A UserChangesMap` instead of `The UserChangesMap` because you have one for each set of changes

interim17 committed:
define prevState const in handler tests

interim17 committed:
pass seletions down as prop

interim17 committed:
add default array for selections
